src/libopensc/card-muscle.c

```diff
@@ -475,13 +465,6 @@ static int muscle_init(sc_card_t *card)
 	card->flags |= SC_CARD_FLAG_RNG;
 	card->caps |= SC_CARD_CAP_RNG;
 
-	/* Card type detection */
-	if (_sc_match_atr(card, muscle_atrs, &card->type) < 0)   {
```

Comment:
You may move the the ATR detection to `muscle_match_card()`, but keep muscle limited to these ATRs by default. The user may still force the muscle driver via `opensc.conf`.

Reply:
But if the card does not match with ``` muscle_atrs ``` it should return not supported? Because i forced the driver in ``` opensc.conf ``` but in the ATR detection does not match.

Reply:
If the ATR detection is in `muscle_match_card()` instead of `muscle_init()` you should be able to force the driver like this:
```
card_atr 11:22:33:44:55:66:77:88:99:00 {
	driver = "muscle";
}
```
Alternatively, you may add your ATR to the list...

Reply:
As a general rule, publicly available JavaCard applet based drivers should contain NO hardcoded ATR-s. At least no generic JavaCard ATR-s. ATR-s could be considered IFF the applet sets historical bytes to some pre-defined value (and the given card does not have different cold/warm ATR-s).